<!DOCTYPE html>
<html lang="en">
<head>
   <meta charset="UTF-8">
   <meta name="viewport" content="width=device-width, initial-scale=1.0">
   <title></title>
</head>
<body>
   <script>
      /*las promesas son valores que no conosemos, es la promesa que hay va a ver un valor cuando una accion 
      asincrona suceda y se resuelva. las promesas tienen 3 estados y son como cualquier objeto en javascript:
      
      1.el primer estado y en el cual cualquier promesa se inicializa es el estado de PENDING 
      2.Si la promesa se resuelve exitosamente pasa al esta de FULFILLED
         para obtener el valor de la resolucion de la promesa usamos la funcion   .then(val => ..)
      3.SI ocuerre algun erro en la accion asincrona pasa al estado de REJECTED
         si no se puede resolver la promesa usamos .catch(err => ..)

      como se crea una promesa ?

         new promise(function(resolve.reject){
            ....
         }).then(valor =>{     --> para obtener el valor de la resulucion exitosa
            ..
         }).catch(err => {     --> si sucede un error llamamos a esta funcion 

         })

      cuando una promesa llega al estado fulfilled podemos retornar otra promesa e irlas encadenando y cada
      una de ellas puede ser rechazada o resulta en una nueva promesaa
      
      PERO las promesas tambien pueden resolverse de manera sincrona  
      */

   </script>
     <script src="https://code.jquery.com/jquery-3.5.1.min.js"
     integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0=" crossorigin="anonymous"></script>
  <script>
 

     const API_URL = 'https://swapi.dev/api/'
     const PEOPLE_URL = 'people/:id'

     const opts = { crossDomain: true }



    function obtenerPersonaje(id){
       return new Promise ((resolve, rejec) =>{
          const url = `${API_URL}${PEOPLE_URL.replace(':id', id)}`
          $
            .get(url, opts, function(data){
               resolve(data)
            })
            .fail(() => rejec(id))
       })
    }

    function onError (id){
       console.log(`Error al obtener el personaje ${id}  `)
    }

    obtenerPersonaje(1)
      .then(function(personaje){
         console.log(`El personaje 1 es ${personaje.name}`)
      })
      .catch(function(id){
         onError(id)
      })
      
   </script>
</body>
</html>